---
title: "RNASeqDGE"
author: "@DoubleOmics"
runtime: shiny
output:
    html_document:
        css: my.css
---

```{r, echo=FALSE, comments="", warning=FALSE, message=FALSE}

rm(list=ls())
library(tidyverse)
library(readxl)
library(shiny)
library(ggplot2)
library(DESeq2)
library(reshape2)
library(calibrate)
options(stringsAsFactors=FALSE)

##Local func
getres<-function(countfile, phenofile, comparison, level1, level0, AdjustedCutoff=0.05, FCCutoff=1){
  #read files
  count <- read_excel(countfile)%>%
    column_to_rownames(var="Gene")
  pheno <- read_excel(phenofile)%>%
    column_to_rownames(var="Samples")
  
  if (!identical(colnames(count), rownames(pheno))) {
    message("Sample names in count matrix must be identical to the sample names in pheno data")
    break
  } 
  
  #dds object
  dds <- DESeqDataSetFromMatrix(countData=count, 
                                colData=pheno, 
                                design=as.formula(paste0("~", comparison)))
  dds<-dds[rowSums(counts(dds))>0, ]
  
  normalized_dat<-assay(rlog(dds, blind=TRUE))%>%data.frame%>%rownames_to_column(var="Gene")
  dds<-DESeq(dds)
  
  #results
  res<-results(dds, contrast=c(comparison, level1, level0))%>%data.frame%>%
    rownames_to_column(var="Gene")%>%
    filter(complete.cases(.))%>%
    mutate(log10padj=-log10(padj))%>%
    arrange(pvalue)
  res<-res%>%
    mutate(Significance=ifelse(log2FoldChange < (-FCCutoff) & padj < AdjustedCutoff, "FC_FDR_Down",
                               ifelse(log2FoldChange > FCCutoff & padj < AdjustedCutoff, "FC_FDR_Up", "NS")))
  
  #plot
  p <-ggplot(res, aes(x=log2FoldChange, y=log10padj)) +
    geom_point(aes(color=factor(Significance)), alpha=1/2, size=2) +
    theme_bw(base_size=16) +
    xlab(bquote(~Log[2]~ "fold change")) +
    ylab(bquote(~-Log[10]~adjusted~italic(P))) +
    geom_vline(xintercept=c(-FCCutoff, FCCutoff), linetype="longdash", colour="black", size=0.4) +
    geom_hline(yintercept=-log10(AdjustedCutoff), linetype="longdash", colour="black", size=0.4)
  
  return(list(results=res,
              normal=normalized_dat,
              plot=p))
}

# Define UI ----
ui <- fluidPage(
  tabsetPanel(
    #tabPanel-Input
    tabPanel("Input", fluid = TRUE,
             
             # tab title ----
             titlePanel("Upload data"),
             
             # sidebar layout with input and output tables ----
             sidebarLayout(
               
               # sidebar panel for inputs ----
               sidebarPanel(
                 
                 #example1 link
                 uiOutput("mexample"),
                 
                 # input1: Select a file ----
                 fileInput("file1", "Count matrix File (.xlsx)",
                           multiple = TRUE,
                           accept = c("text/csv",
                                      "text/comma-separated-values,text/plain",
                                      ".csv")),

                 #example2 link
                 uiOutput("pexample"),
                 
                 #input2: select a file ----
                 fileInput("file2", "Manifest File (.xlsx)",
                           multiple = TRUE,
                           accept = c("text/csv",
                                      "text/comma-separated-values,text/plain",
                                      ".csv")),
                 
                 #select column name
                 selectInput("design", "Column name for analysis", " "),
                 
                 #select ref group
                 uiOutput("level0"),
                 
                 #select study group
                 uiOutput("level1"),
                 
                 #action run
                 actionButton("runbutton", "Run"),
                 
                 #comment message
                 p("Click to perform differential gene expression analysis between the selected groups"),
                 
                 #README link
                 uiOutput("README"),
                 
                 #issue report
                 uiOutput("issue")
                 
               ),
               # Main panel for displaying outputs ----
               mainPanel(
                 
                 # Output: Data file ----
                 tableOutput("matrix"),
                 tableOutput("pdat")
               )
             )
    ),
    
    #tabPanel-Results
    tabPanel("Results", fluid = TRUE,
             # App title ----
             titlePanel("Download results"),
             
             # Sidebar layout with input and output definitions ----
             sidebarLayout(
               
               # Sidebar panel for inputs ----
               sidebarPanel(
                 
                 # Input: Choose dataset ----
                 selectInput("results", "Choose a dataset:",
                             choices = c("Results", "Normalized matrix")),
                 
                 # Button
                 downloadButton("downloadData", "Download")
                 
               ),
               
               # Main panel for displaying outputs ----
               mainPanel(
                 
                 tableOutput("table")
                 
               )
               
             )             
             
    ),
    #tabPanel-Plots
    tabPanel("Plots", fluid = TRUE,
             fluidRow(
               column(width = 8,
                      plotOutput("plot1", height = 800,
                                 # Equivalent to: click = clickOpts(id = "plot_click")
                                 click = "plot1_click",
                                 brush = brushOpts(
                                   id = "plot1_brush"
                                 )
                      )
               ),
               column(width = 4,
                      h4("Brushed points"),
                      verbatimTextOutput("brush_info")
               )
             )
    )
  )
)


# Define Server ----
server <- function(input, output, session) {
  
  #tabPanel-Input
  datobj <- reactive({
    req(input$file1)
    req(input$file2)
    count <- read_excel(input$file1$datapath)
    pheno <- read_excel(input$file2$datapath)
    return(list(counts=count, 
                pheno=pheno))
  })
  
  
  ### matrix file

  output$matrix <- renderTable({
      return(head(datobj()[["counts"]]))
  })
  
  url1 <- a("Count matrix example", href="https://github.com/DadongZ/RNASeqDGE/tree/master/data")

  output$mexample <- renderUI({
    tagList(url1)
  })
  
  output$pdat <- renderTable({
      return(datobj()[["pheno"]])
  })
  
  ### pheno file 
  url2 <- a("Manifest example", href="https://github.com/DadongZ/RNASeqDGE/tree/master/data")
  
  output$pexample <- renderUI({
    tagList( url2)
  })

  observe({
    updateSelectInput(session, "design", choices=names(datobj()[["pheno"]]))
  })
  
  output$level0 <- renderUI({
     selectInput("ref0", "Reference group", datobj()[["pheno"]][[input$design]])
  })

  
  output$level1 <- renderUI({
     selectInput("ref1", "Study group", datobj()[["pheno"]][[input$design]])
  })

  
  readme <- a("README", href="https://github.com/DadongZ/RNASeqDGE/blob/master/README.md")
  output$README <- renderUI({
    tagList("Need help? ", readme)
  })
  
  issue <- a("issues", href="https://github.com/DadongZ/RNASeqDGE/issues")
  output$issue <- renderUI({
    tagList("Please report issues at: ", issue)
  })

  #tabPanel-Results
  resobj <- eventReactive(input$runbutton, {
    res<-getres(input$file1$datapath, input$file2$datapath,
                comparison=input$design, 
                level1=input$ref1,
                level0=input$ref0)
    return(list(normal=res[["normal"]],
                results=res[["results"]],
                volcano=res[["plot"]]))
  })
  
  todowndat <- reactive({
    switch(input$results,
           "Results" = resobj()[["results"]],
           "Normalized matrix" = resobj()[["normal"]]
    )
  })
  
  output$table <- renderTable({
    todowndat()
  })
  
  output$downloadData <- downloadHandler(
    filename = function() {
      paste(input$results, ".csv", sep = "")
    },
    content = function(file) {
      write.csv(todowndat(), file, row.names = FALSE)
    }
  ) 
  
  #tabPanel-Results
  output$plot1 <- renderPlot({
    resobj()[["volcano"]]
  })
  
  output$brush_info <- renderPrint({
    showdf<-resobj()[["results"]]%>%dplyr::select(Gene, log2FoldChange, pvalue, padj, log10padj) 
    brushedPoints(showdf, input$plot1_brush)
  })
}


## Run the app
shinyApp(ui, server, options = list(height = 1100))

```

