---
title: "RNASeqDGE"
author: "@DoubleOmics"
runtime: shiny
output:
    html_document:
        css: my.css
---

```{r, echo=FALSE, comments="", warning=FALSE, message=FALSE}
rm(list=ls())
library(tidyverse)
library(readxl)
library(shiny)
library(shinyWidgets)
library(ggplot2)
library(DESeq2)
library(reshape2)
library(calibrate)
library(airway)
library(gage)
data(airway)
options(stringsAsFactors=FALSE)

##Local func
getres<-function(count, pdat, comparison, level1, level0, AdjustedCutoff=0.05, FCCutoff=1){
  gcol<-colnames(count)[1]
  count<-count%>%data.frame%>%
    column_to_rownames(var=gcol)
  pcol<-colnames(pdat)[1]
  pdat<-pdat%>%data.frame%>%
    column_to_rownames(var=pcol)
    
  if (!identical(colnames(count), rownames(pdat))) {
    message("Sample names in count matrix must be identical to the sample names in pheno data")
    break
  } 
  
  #dds object
  dds <- DESeqDataSetFromMatrix(countData=count, 
                                colData=pdat, 
                                design=as.formula(paste0("~", comparison)))
  dds<-dds[rowSums(counts(dds))>0, ]
  
  normalized_dat<-assay(rlog(dds, blind=TRUE))%>%data.frame%>%rownames_to_column(var="Gene")
  dds<-DESeq(dds)
  
  #results
  res<-results(dds, contrast=c(comparison, level1, level0))%>%data.frame%>%
    rownames_to_column(var="Gene")%>%
    filter(complete.cases(.))%>%
    mutate(log10padj=-log10(padj))%>%
    arrange(pvalue)
  res<-res%>%
    mutate(Significance=ifelse(log2FoldChange < (-FCCutoff) & padj < AdjustedCutoff, "FC_FDR_Down",
                               ifelse(log2FoldChange > FCCutoff & padj < AdjustedCutoff, "FC_FDR_Up", "NS")))
  
  #plot
  p <-ggplot(res, aes(x=log2FoldChange, y=log10padj)) +
    geom_point(aes(color=factor(Significance)), alpha=1/2, size=2) +
    theme_bw(base_size=16) +
    xlab(bquote(~Log[2]~ "fold change")) +
    ylab(bquote(~-Log[10]~adjusted~italic(P))) +
    geom_vline(xintercept=c(-FCCutoff, FCCutoff), linetype="longdash", colour="black", size=0.4) +
    geom_hline(yintercept=-log10(AdjustedCutoff), linetype="longdash", colour="black", size=0.4)
  
  return(list(results=res,
              normal=normalized_dat,
              plot=p))
}

# Define UI ----
ui <- fluidPage(
  tabsetPanel(
    #tabPanel-Input
    tabPanel("Input", fluid = TRUE,
             
             # tab title ----
             titlePanel("Upload data"),
             
             # sidebar layout with input and output tables ----
             sidebarLayout(
               
               # sidebar panel for inputs ----
               sidebarPanel(
                 
                 #show ct demo
                 actionBttn("runexample", "RUN demo data", style="simple", size="sm", color = "primary"),
                 
                 # input1: Select a file ----
                 fileInput("file1", "Count matrix File (.xlsx)",
                           multiple = TRUE,
                           accept = c("text/csv",
                                      "text/comma-separated-values,text/plain",
                                      ".csv")),
                 
                 #input2: select a file ----
                 fileInput("file2", "Manifest File (.xlsx)",
                           multiple = TRUE,
                           accept = c("text/csv",
                                      "text/comma-separated-values,text/plain",
                                      ".csv")),
                 
                 #select column name
                 selectInput("design", "Column name for analysis", " "),
                 
                 #select ref group
                 uiOutput("level0"),
                 
                 #select study group
                 uiOutput("level1"),
                 
                 #action run
                 actionBttn("runbutton", "GO", style="simple", size="sm", color = "primary"),
                 
                 actionBttn("reset", "RESET", style="simple", size="sm", color = "warning"),
                 #comment message
                 p("Click GO to perform differential gene expression analysis between the selected groups"),
                 
                 #README link
                 uiOutput("README"),
                 
                 #issue report
                 uiOutput("issue")
                 
               ),
               # Main panel for displaying outputs ----
               mainPanel(
                 
                 # Output: Data file ----
                 tableOutput("matrix"),
                 tableOutput("pdat")
               )
             )
    ),
    
    #tabPanel-Results
    tabPanel("Results", fluid = TRUE,
             # App title ----
             titlePanel("Download results"),
             
             # Sidebar layout with input and output definitions ----
             sidebarLayout(
               
               # Sidebar panel for inputs ----
               sidebarPanel(
                 
                 # Input: Choose dataset ----
                 selectInput("results", "Choose a dataset:",
                             choices = c("Results", "Normalized matrix")),
                 
                 # Button
                 downloadButton("downloadData", "Download")
                 
               ),
               
               # Main panel for displaying outputs ----
               mainPanel(
                 
                 tableOutput("table")
                 
               )
               
             )             
             
    ),
    #tabPanel-Plots
    tabPanel("Plots", fluid = TRUE,
             fluidRow(
               column(width = 8,
                      plotOutput("plot1", height = 800,
                                 # Equivalent to: click = clickOpts(id = "plot_click")
                                 click = "plot1_click",
                                 brush = brushOpts(
                                   id = "plot1_brush"
                                 )
                      )
               ),
               column(width = 4,
                      h4("Brushed points"),
                      verbatimTextOutput("brush_info")
               )
             )
    )
  )
)


# Define Server ----
server <- function(input, output, session) {
  
  #tabPanel-Input
  
  ###demo data
  ####count
  set.seed(123)
  ctdemo<- assay(airway, "counts")
  ctdemo<-ctdemo[sample(1:nrow(ctdemo), 2000), ]%>%data.frame()%>%
    rownames_to_column(var="Gene")
  ####manifest
  pdemo<-colData(airway)%>%data.frame%>%
    rownames_to_column(var="ID")
  ###display demo count matrix
  observeEvent(input$runexample, {
    output$matrix <- renderTable({
      head(ctdemo, 10)
    })
    output$pdat <- renderTable({
      head(pdemo, 10)
    })
    
    observe({
      updateSelectInput(session, "design", choices="dex")
    })
    
    output$level0 <- renderUI({
      selectInput("ref0", "Reference group", "untrt")
    })
    
    output$level1 <- renderUI({
      selectInput("ref1", "Study group", "trt")
    })
    
    resdemo <- reactiveVal()
    observeEvent(input$runbutton, {
      resdemo <- (NULL)
      withProgress(message = 'Calculation in progress', style="old", value=0, {
        Sys.sleep(1)
        resdemo(getres(ctdemo, pdemo,
                       comparison="dex", 
                       level1="trt",
                       level0="untrt"))
        incProgress(1)
      })
    })
    todowndat <- reactive({
      switch(input$results,
             "Results" = resdemo()[["results"]],
             "Normalized matrix" = resdemo()[["normal"]]
      )
    })
    
    output$table <- renderTable({
      todowndat()
    })
    
    output$downloadData <- downloadHandler(
      filename = function() {
        paste(input$results, ".csv", sep = "")
      },
      content = function(file) {
        write.csv(todowndat(), file, row.names = FALSE)
      }
    ) 
    #tabPanel-Results
    output$plot1 <- renderPlot({
      resdemo()[["plot"]]
    })
    
    output$brush_info <- renderPrint({
      showdf<-resdemo()[["results"]]%>%dplyr::select(Gene, log2FoldChange, pvalue, padj, log10padj) 
      brushedPoints(showdf, input$plot1_brush)
    })
  })
  
  observeEvent(input$reset, {
    
    ctobj <- reactive({
      req(input$file1)
      count <- read_excel(input$file1$datapath)
      return(count)
    })
    
    pobj <- reactive({
      req(input$file2)
      pheno <- read_excel(input$file2$datapath)
      return(pheno=pheno)
    })
    
    
    ### matrix file
    output$matrix <- renderTable({
      head(ctobj())
    })
    
    output$pdat <- renderTable({
      pobj()
    })
    
    ### pheno file 
    observe({
      updateSelectInput(session, "design", choices=names(pobj()))
    })
    
    output$level0 <- renderUI({
      selectInput("ref0", "Reference group", pobj()[[input$design]])
    })
    
    output$level1 <- renderUI({
      selectInput("ref1", "Study group", pobj()[[input$design]])
    })
    

      resobj <- reactiveVal()
      observeEvent(input$runbutton, {
        resobj <- (NULL)
        withProgress(message = 'Calculation in progress', style="old", value=0, {
          Sys.sleep(1)
          resobj(getres(ctobj(), pobj(),
                        comparison=input$design, 
                        level1=input$ref1,
                        level0=input$ref0))
          incProgress(1)
        })
      })
      
      todowndat <- reactive({
        switch(input$results,
               "Results" = resobj()[["results"]],
               "Normalized matrix" = resobj()[["normal"]]
        )
      })
      
      output$table <- renderTable({
        todowndat()
      })
      
      output$downloadData <- downloadHandler(
        filename = function() {
          paste(input$results, ".csv", sep = "")
        },
        content = function(file) {
          write.csv(todowndat(), file, row.names = FALSE)
        }
      ) 
      
      #tabPanel-Results
      output$plot1 <- renderPlot({
        resobj()[["plot"]]
      })
      
      output$brush_info <- renderPrint({
        showdf<-resobj()[["results"]]%>%dplyr::select(Gene, log2FoldChange, pvalue, padj, log10padj) 
        brushedPoints(showdf, input$plot1_brush)
      })
   
  })
  
  readme <- a("README", href="https://github.com/DadongZ/RNASeqDGE/blob/master/README.md")
  output$README <- renderUI({
    tagList("Need help? ", readme)
  })
  
  issue <- a("issues", href="https://github.com/DadongZ/RNASeqDGE/issues")
  output$issue <- renderUI({
    tagList("Please report issues at: ", issue)
  })
  
}

## Run the app
shinyApp(ui, server, options = list(height = 1100))

```